toolchain/base/yaml.h

```diff
@@ -0,0 +1,102 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_YAML_H_
+#define CARBON_TOOLCHAIN_BASE_YAML_H_
+
+#include "common/check.h"
+#include "common/ostream.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/YAMLTraits.h"
+
+// This file provides adapters for outputting YAML using llvm::yaml's APIs. It
+// only supports output, not input. However, it addresses the mix of const and
+// non-const expectations of the llvm::yaml that make it difficult to otherwise
+// use the trait-based approach.
+
+namespace Carbon::Yaml {
+
+// Helper for printing YAML, to maintain a consistent configuration.
+template <typename T>
+inline auto Print(llvm::raw_ostream& out, T yaml) -> void {
+  llvm::yaml::Output yout(out, /*Ctxt=*/nullptr, /*WrapColumn=*/80);
+  yout << yaml;
+}
+
+// Similar to the standard Printable<T>, but relies on OutputYaml for printing.
+template <typename T>
+class Printable : public Carbon::Printable<T> {
+ public:
+  auto Print(llvm::raw_ostream& out) const -> void {
+    Carbon::Yaml::Print(out, static_cast<const T*>(this)->OutputYaml());
+  }
+};
+
+// Adapts a function for outputting YAML as a scalar. This currently assumes no
+// scalars passed through this should be quoted.
+class OutputScalar {
+ public:
+  explicit OutputScalar(std::function<void(llvm::raw_ostream&)> output)
+      : output_(std::move(output)) {}
+
+  auto Output(llvm::raw_ostream& out) const -> void { output_(out); }
+
+ private:
+  std::function<void(llvm::raw_ostream&)> output_;
+};
+
+// Adapts a function for outputting YAML as a mapping.
+class OutputMapping {
+ public:
+  // Maps a value. This mainly takes responsibility for copying the value,
+  // letting mapRequired take `&value`.
+  template <typename T>
+  static auto Map(llvm::yaml::IO& io, llvm::StringRef key, T value) -> void {
+    io.mapRequired(key.data(), value);
+  }
+  static auto Map(llvm::yaml::IO& io, llvm::StringRef key,
+                  const llvm::StringRef value) -> void {
+    llvm::StringRef v = value;
+    io.mapRequired(key.data(), v);
```

Comment:
```suggestion
                  llvm::StringRef value) -> void {
    io.mapRequired(key.data(), value);
```

Reply:
Done, this must've been a leftover somehow.